### PR TITLE
Cql4bonnie use mat artifacts

### PR DESCRIPTION
**Pull request by jbradl11**
Pull requests into Bonnie Bundler require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

This PR is to implement the ability to use the MAT provided JSON and XML artifacts, rather than generating our own. We also now store the MAT Package in a separate collection. 

Related JIRA Tickets: 
- https://jira.mitre.org/browse/BONNIE-1143
- https://jira.mitre.org/browse/BONNIE-1145
- https://jira.mitre.org/browse/BONNIE-959

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] JIRA ticket for this PR:
- [x] JIRA ticket links to this PR
- [x] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] If UI changes have been made, google WAVE plug-in has been executed to ensure no 508 issues were introduced.
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [x] Code coverage has not gone down and all code touched or added is covered.
    * In rare situations, this may not be possible or applicable to a PR. In those situations:
        1. Note why this could not be done or is not applicable here:
        2. Add TODOs in the code noting that it requires a test
        3. Add a JIRA task to add the test and link it here:
- [x] Automated regression test(s) pass

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA test links:
- [x] Justification for using JIRA tests:
- [x] JIRA tests have been added to sprint

**Reviewer 1:** @holmesie

Name:
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
N/A - No JIRA tests
- [x] JIRA tests have been run and pass N/A
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree N/A

**Reviewer 2:**

Name: @losborne
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA tests have been run and pass (N/A)
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree (N/A)

### CODE REVIEW

File: lib/measures/loading/cql_loader.rb [hunk 1]
```diff
@@ -254,151 +260,52 @@ def self.generate_single_code_references(elms, all_codes_and_code_names, user)
...
-          end
-        end
-      end
-      return cql_libraries, model
-    end
-
-    # Checks if the given string is a reserved keyword in JavaScript. Useful
-    # for sanitizing potential user input from imported CQL code.
-    def self.is_javascript_keyword(string)
-      ['do', 'if', 'in', 'for', 'let', 'new', 'try', 'var', 'case', 'else', 'enum', 'eval', 'false', 'null', 'this', 'true', 'void', 'with', 'break', 'catch', 'class', 'const', 'super', 'throw', 'while', 'yield', 'delete', 'export', 'import', 'public', 'return', 'static', 'switch', 'typeof', 'default', 'extends', 'finally', 'package', 'private', 'continue', 'debugger', 'function', 'arguments', 'interface', 'protected', 'implements', 'instanceof'].include? string
```
Comment (holmesie) [inline lib/measures/loading/cql_loader.rb:382]: Removing the `remove_spaces_in_functions` makes sense because you're handling spaces differently in the execution engine, but do we also now not care about this? I'm pretty sure the answer to that is we no longer need it, but I want to be absolutely sure we don't have things like `eval` popping up in code that is executed.

**Review by holmesie — CHANGES_REQUESTED** [review]

File: lib/measures/loading/cql_loader.rb (hunk 1 above)
Comment (jbradl11) [reply lib/measures/loading/cql_loader.rb:382]: All references to `eval` in the execution engine have been removed. Once the cql_qdm_patient_api PR gets merged in (which should happen before this gets merged), we will no longer be relying on it.

File: lib/measures/cql_to_elm_helper.rb [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+          # The error text will be written to a load_error file and will not be displayed in the error dialog displayed to the user since
+          # measures_controller.rb does not handle this type of exception
+          raise MeasureLoadingException.new "Error Translating CQL to ELM: " + errorMsg
+        end
+      end
+
+      private
+      # Parse the JSON response into an array of json objects (one for each library)
+      def self.parse_elm_response(response)
+        # Not the same delimiter in the response as we specify ourselves in the request,
```
Comment (lukoktonos) [inline lib/measures/cql_to_elm_helper.rb]: Lines 54-61 and 68-75 are duplicated and should be a helper function.

File: lib/measures/cql_to_elm_helper.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        delimiter = response.split("\r\n")[0].strip
+        parts = response.split(delimiter)
+        # The first part will always be an empty string. Just remove it.
+        parts.shift
+        # The last part will be the "--". Just remove it.
+        parts.pop
+  
+        parsed_parts = []
+        parts.each do |part|
+          lines = part.split("\r\n")
```
Comment (losborne) [inline lib/measures/cql_to_elm_helper.rb]: These 3 lines are also duplicated and could be put into a separate helper function that the above requested helper function could call.

File: lib/measures/loading/cql_loader.rb [hunk 4]
```diff
@@ -68,10 +66,15 @@ def self.load_mat_cql_exports(user, zip_file, out_dir, measure_details, vsac_use
...
-      measure = Measures::Loader.load_hqmf_cql_model_json(json, user, cql_artifacts[:all_value_set_oids], main_cql_library, cql_artifacts[:cql_definition_dependency_structure],
-                                                          cql_artifacts[:elms], cql_artifacts[:elm_annotations], cql_libraries, nil, cql_artifacts[:value_set_oid_version_objects])
+      measure = Measures::Loader.load_hqmf_cql_model_json(json, user, cql_artifacts[:all_value_set_oids], main_cql_library, cql_artifacts[:cql_definition_dependency_structure], 
+                                                          cql_artifacts[:elms], cql_artifacts[:elm_annotations], files[:CQL], nil, cql_artifacts[:value_set_oid_version_objects])
       measure['episode_of_care'] = measure_details['episode_of_care']
       measure['type'] = measure_details['type']
+
+      # Create, associate and save the measure package.
+      measure.package = CqlMeasurePackage.new(file: BSON::Binary.new(zip_file.read()))
+      measure.package.save
```
Comment (lukoktonos) [inline lib/measures/loading/cql_loader.rb:76]: Should this and the below calls to save be `measure.package.save!` to throw an exception if it fails?  There isn't a check to see if save returns true or false.  I'm not sure if it matters in this case or not.

**Review by lukoktonos — CHANGES_REQUESTED** [review]

File: lib/measures/loading/cql_loader.rb (hunk 4 above)
Comment (holmesie) [reply lib/measures/loading/cql_loader.rb:76]: I think that just fails if any validations fail, which I'm pretty sure our measure model doesn't have.

**Review by holmesie — DISMISSED** [review]
Tests pass, code looks good. Approving, with the understanding that Luke's remaining comments are addressed as well.

File: lib/measures/loading/cql_loader.rb (hunk 4 above)
Comment (hossenlopp) [reply lib/measures/loading/cql_loader.rb:76]: I'm indifferent on the `save` or `save!` of this. The only validation that may happen is if it is indeed a binary type for the file field. Which that is specifically being done in the above line. We have already checked to make sure it is a valid package much earlier, so we shouldn't have to worry about that either.

If it fails to save the package, it's not a huge loss to us, Bonnie will still function. But if we do the misnamed `rebuild_elm` rake task then we won't have the original package to reprocess the elm.

What do you guys think I should do?

Comment (lukoktonos) [reply lib/measures/loading/cql_loader.rb:76]: Sounds like leaving it as save is fine.

Comment (holmesie) [reply lib/measures/loading/cql_loader.rb:76]: I'd say leave it, I think your justification makes perfect sense.

**Review by lukoktonos — APPROVED** [review]
Changes look good--approved.

**Review by holmesie — APPROVED** [review]
Approved.
